### PR TITLE
Add/robot arduino controller

### PR DESCRIPTION
**Pull request by Tyson453**

### CODE REVIEW

File: robotbase/src/main/java/com/team2357/frc2022/arduino/RobotArduino.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+
+    public RobotArduino(String ttyDevice) {
+        super(ttyDevice);
+        super.start();
+    }
+
+    public boolean getIntakeValue() {
+        boolean value = false;
+        System.out.println(super.isConnected());
```
Comment (NolanCampbell2357) [inline robotbase/src/main/java/com/team2357/frc2022/arduino/RobotArduino.java:16]: We can remove this line

File: robotbase/src/main/java/com/team2357/lib/arduino/ArduinoUSBController.java [hunk 2]
```diff
@@ -40,9 +40,14 @@
...
-		m_serialPort = SerialPort.getCommPort(ttyDevice);
-		m_serialPort.setComPortParameters(115200, 8, SerialPort.ONE_STOP_BIT, SerialPort.NO_PARITY);
-		m_serialPort.setComPortTimeouts(SerialPort.TIMEOUT_READ_SEMI_BLOCKING, READ_TIMEOUT, 0);
+		try {
+			m_serialPort = SerialPort.getCommPort(ttyDevice);
+			m_serialPort.setComPortParameters(115200, 8, SerialPort.ONE_STOP_BIT, SerialPort.NO_PARITY);
+			m_serialPort.setComPortTimeouts(SerialPort.TIMEOUT_READ_SEMI_BLOCKING, READ_TIMEOUT, 0);
+		} catch (Exception e) {
+			m_serialPort = null;
+			System.out.println(e.getMessage());
```
Comment (NolanCampbell2357) [inline robotbase/src/main/java/com/team2357/lib/arduino/ArduinoUSBController.java:49]: Should be ```System.err.println()```.

File: robotbase/src/main/java/com/team2357/frc2022/RobotContainer.java [hunk 3]
```diff
@@ -35,15 +37,30 @@
   private final IntakeDriveControls m_driverControls;
   private final GunnerControls m_gunnerControls;
 
+  private final RobotArduino m_arduinoSensor;
+
+
```
Comment (NolanCampbell2357) [inline robotbase/src/main/java/com/team2357/frc2022/RobotContainer.java:42]: Typically we only have one blank line between sets of variables, or between functions. Not a big deal, just for your information.

File: robotbase/src/main/java/com/team2357/frc2022/RobotContainer.java [hunk 4]
```diff
@@ -35,15 +37,30 @@
...
+    ArduinoSensorState intakeIRSensor = () -> {
+      return m_arduinoSensor.getIntakeValue();
+    };
+    ArduinoSensorState feederIRSensor = () -> {
+      return m_arduinoSensor.getFeederValue();
+    };
+    ArduinoSensorState turretIRSensor = () -> {
+      return m_arduinoSensor.getTurretValue();
+    };
+
```
Comment (NolanCampbell2357) [inline robotbase/src/main/java/com/team2357/frc2022/RobotContainer.java:57]: See above, excess blank line here as well.

File: robotbase/src/main/java/com/team2357/frc2022/arduino/RobotArduino.java [hunk 5]
```diff
@@ -0,0 +1,42 @@
+package com.team2357.frc2022.arduino;
+
+import com.team2357.frc2022.Constants;
+import com.team2357.lib.arduino.ArduinoUSBController;
+
+public class RobotArduino extends ArduinoUSBController{
+
+
```
Comment (NolanCampbell2357) [inline robotbase/src/main/java/com/team2357/frc2022/arduino/RobotArduino.java:8]: Extra line.

**Review by NolanCampbell2357 — CHANGES_REQUESTED** [review]
Almost ready, just a couple formatting things we did not catch Yesterday.

File: robotbase/src/main/java/com/team2357/frc2022/arduino/ArduinoSensorState.java [hunk 6]
```diff
@@ -0,0 +1,6 @@
+package com.team2357.frc2022.arduino;
+
+@FunctionalInterface
+public interface ArduinoSensorState {
```
Comment (coderkevin) [inline robotbase/src/main/java/com/team2357/frc2022/arduino/ArduinoSensorState.java:4]: This should be more specific. Not all "SensorStates" will be true/false. Additionally, not all sensor states will be coming from an Arduino. I'd rename this to `SensorBooleanState` to be more descriptive.

Comment (coderkevin) [reply robotbase/src/main/java/com/team2357/frc2022/arduino/ArduinoSensorState.java:4]: I'd also move this from the `arduino` package to somewhere more general. Keep in mind, we could possibly move one of our sensors over to the RoboRIO DIO itself, so whatever we do here needs to make sense in that context as well.

File: robotbase/src/main/java/com/team2357/frc2022/arduino/RobotArduino.java [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+    public RobotArduino(String ttyDevice) {
+        super(ttyDevice);
+        super.start();
+    }
+
+    public boolean getIntakeValue() {
+        boolean value = false;
+
+        if (super.isConnected()){
+            value = !super.getDeviceFieldBoolean(Constants.ARDUINO.INTAKE_SENSOR_JSON_NAME, Constants.ARDUINO.INTAKE_SENSOR_STATE_FIELD);
```
Comment (coderkevin) [inline robotbase/src/main/java/com/team2357/frc2022/arduino/RobotArduino.java:17]: Why are we looking for a `false` value here? Shouldn't the arduino be sending 'true' when the intake has a cargo present?

File: robotbase/src/main/java/com/team2357/frc2022/arduino/RobotArduino.java [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+        }
+        
+        return value;
+    }
+
+    public boolean getFeederValue() {
+        boolean value = false;
+        if (super.isConnected()){
+            value = !super.getDeviceFieldBoolean(Constants.ARDUINO.FEEDER_SENSOR_JSON_NAME, Constants.ARDUINO.FEEDER_SENSOR_STATE_FIELD);
```
Comment (coderkevin) [inline robotbase/src/main/java/com/team2357/frc2022/arduino/RobotArduino.java:26]: Same ? as above

File: robotbase/src/main/java/com/team2357/frc2022/arduino/RobotArduino.java [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+        }
+        
+        return value;
+    }
+
+    public boolean getTurretValue() {
+        boolean value = false;
+        if (super.isConnected()){
+            value = !super.getDeviceFieldBoolean(Constants.ARDUINO.TURRET_SENSOR_JSON_NAME, Constants.ARDUINO.TURRET_SENSOR_STATE_FIELD);
```
Comment (coderkevin) [inline robotbase/src/main/java/com/team2357/frc2022/arduino/RobotArduino.java:35]: Same ? as above

File: robotbase/src/main/java/com/team2357/frc2022/subsystems/FeederSubsystem.java [hunk 10]
```diff
@@ -1,21 +1,20 @@
...
+import com.team2357.frc2022.arduino.ArduinoSensorState;
 import com.team2357.lib.arduino.ArduinoUSBController;
 import com.team2357.lib.subsystems.ClosedLoopSubsystem;
 import com.ctre.phoenix.motorcontrol.ControlMode;
 import com.ctre.phoenix.motorcontrol.can.WPI_TalonSRX;
 
 public class FeederSubsystem extends ClosedLoopSubsystem {
     private WPI_TalonSRX m_feederMotor;
-    private ArduinoUSBController m_arduinoIRSensor;
+    private ArduinoSensorState m_isCargoAtFeederWheel;
```
Comment (coderkevin) [inline robotbase/src/main/java/com/team2357/frc2022/subsystems/FeederSubsystem.java:12]: The variable here isn't really a "getter" function itself. I'd probably name it something different like "m_feederSensor" so when you call it below it's `m_cargoFeeder.getState()` which makes more sense to me.

File: robotbase/src/main/java/com/team2357/frc2022/subsystems/IntakeSubsystem.java [hunk 11]
```diff
@@ -16,13 +17,15 @@
...
+    private ArduinoSensorState m_isCargoInIntake;
 
     /**
      * @param intakeVictor Victor SPX to use to control intake
      */
-    public IntakeSubsystem(VictorSPX intakeVictor, DoubleSolenoid intakeSolenoid) {
+    public IntakeSubsystem(VictorSPX intakeVictor, DoubleSolenoid intakeSolenoid, ArduinoSensorState isCargoInIntake) {
         m_intakeSolenoid = intakeSolenoid;
         m_intakeSolenoid.set(Value.kOff);
+        m_isCargoInIntake = isCargoInIntake;
```
Comment (coderkevin) [inline robotbase/src/main/java/com/team2357/frc2022/subsystems/IntakeSubsystem.java:28]: Same here, the object isn't a getter itself.

File: robotbase/src/main/java/com/team2357/frc2022/subsystems/SubsystemFactory.java [hunk 12]
```diff
@@ -44,16 +45,16 @@ public FalconTrajectoryDriveSubsystem CreateFalconTrajectoryDriveSubsystem() {
         return subsystem;
     }
 
-    public IntakeSubsystem CreateIntakeSubsystem() {
+    public IntakeSubsystem CreateIntakeSubsystem(ArduinoSensorState state) {
```
Comment (coderkevin) [inline robotbase/src/main/java/com/team2357/frc2022/subsystems/SubsystemFactory.java:48]: Variable name of `state` is not descriptive. Suggest something like `intakeCargoSensor` instead.

File: robotbase/src/main/java/com/team2357/frc2022/subsystems/SubsystemFactory.java [hunk 13]
```diff
@@ -44,16 +45,16 @@ public FalconTrajectoryDriveSubsystem CreateFalconTrajectoryDriveSubsystem() {
...
         DoubleSolenoid intakeDoubleSolenoid = new DoubleSolenoid(PneumaticsModuleType.REVPH,
                 Constants.PCM_ID.INTAKE_SOLENOID_FORWARD_CHANNEL, Constants.PCM_ID.INTAKE_SOLENOID_REVERSE_CHANNEL);
         VictorSPX intakeVictor = new VictorSPX(Constants.CAN_ID.INTAKE_MOTOR_ID);
-        return new IntakeSubsystem(intakeVictor, intakeDoubleSolenoid);
+        return new IntakeSubsystem(intakeVictor, intakeDoubleSolenoid, state);
     }
 
-    public FeederSubsystem CreateFeederSubsystem() {
+    public FeederSubsystem CreateFeederSubsystem(ArduinoSensorState state) {
```
Comment (coderkevin) [inline robotbase/src/main/java/com/team2357/frc2022/subsystems/SubsystemFactory.java:55]: Same as above

File: robotbase/src/main/java/com/team2357/frc2022/arduino/RobotArduino.java [hunk 14]
```diff
@@ -0,0 +1,40 @@
+package com.team2357.frc2022.arduino;
+
+import com.team2357.frc2022.Constants;
+import com.team2357.lib.arduino.ArduinoUSBController;
+
+public class RobotArduino extends ArduinoUSBController{
+
+    public RobotArduino(String ttyDevice) {
+        super(ttyDevice);
+        super.start();
```
Comment (coderkevin) [inline robotbase/src/main/java/com/team2357/frc2022/arduino/RobotArduino.java:10]: I have concerns about starting up a USB serial connection in the constructor here, which is in turn called in the RobotContainer() constructor. We should probably talk about how to ensure the Arduino connection is resilient and test a few scenarios. (like disconnecting and reconnecting the USB cable while it's running, to ensure we can reconnect automatically)

**Review by coderkevin — CHANGES_REQUESTED** [review]

File: robotbase/src/main/java/com/team2357/frc2022/arduino/RobotArduino.java (hunk 14 above)
Comment (NolanCampbell2357) [reply robotbase/src/main/java/com/team2357/frc2022/arduino/RobotArduino.java:11]: I did test to see if the changes made in this PR worked, and we did have USB at this point.

File: robotbase/src/main/java/com/team2357/frc2022/arduino/RobotArduino.java (hunk 7 above)
Comment (NolanCampbell2357) [reply robotbase/src/main/java/com/team2357/frc2022/arduino/RobotArduino.java]: Makes sense, we will make sure this change is reflected on the Arduino side.

**Review by NolanCampbell2357 — APPROVED** [review]
I am satisfied

File: robotbase/src/main/java/com/team2357/lib/arduino/ArduinoUSBController.java [hunk 15]
```diff
@@ -287,6 +288,11 @@ public void stop() {
 	 */
 	@Override
 	public void run() {
+		
+		if (this.isPortOpen()) {
```
Comment (coderkevin) [inline robotbase/src/main/java/com/team2357/lib/arduino/ArduinoUSBController.java:292]: I think this code needs to be removed

File: robotbase/src/main/java/com/team2357/lib/arduino/ArduinoUSBController.java [hunk 16]
```diff
@@ -370,4 +376,8 @@ protected JsonNode getDevice(String deviceName) {
 		}
 		return devices.get(deviceName);
 	}
+
+	public boolean isPortOpen() {
```
Comment (coderkevin) [inline robotbase/src/main/java/com/team2357/lib/arduino/ArduinoUSBController.java:380]: I think this can also be removed

**Review by coderkevin — APPROVED** [review]
Looks good, pre-approving. Please remove the code in ArduinoUSBController before merging.

**Pull request merged**